x-pack/apm-server/aggregation/servicemetrics/aggregator.go

```diff
@@ -228,56 +247,61 @@ type metricsMapEntry struct {
 	aggregationKey
 }
 
-func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) bool {
+func (mb *metricsBuffer) storeOrUpdate(key aggregationKey, metrics serviceMetrics, logger *logp.Logger) {
 	// hash does not use the serviceMetrics so it is safe to call concurrently.
 	hash := key.hash()
 
 	// Full lock because serviceMetrics cannot be updated atomically.
 	mb.mu.Lock()
 	defer mb.mu.Unlock()
 
+	var entry *metricsMapEntry
 	entries, ok := mb.m[hash]
 	if ok {
 		for offset, old := range entries {
 			if old.aggregationKey.equal(key) {
-				entries[offset].serviceMetrics = serviceMetrics{
-					transactionDuration: old.transactionDuration + metrics.transactionDuration,
-					transactionCount:    old.transactionCount + metrics.transactionCount,
-					failureCount:        old.failureCount + metrics.failureCount,
-					successCount:        old.successCount + metrics.successCount,
-				}
-				return true
+				entry = entries[offset]
+				break
 			}
 		}
-	} else if mb.entries >= len(mb.space) {
-		return false
 	}
-
-	half := mb.maxSize / 2
-	if !ok {
-		switch mb.entries {
-		case mb.maxSize:
-			return false
-		case half - 1:
-			logger.Warn("service metrics groups reached 50% capacity")
-		case mb.maxSize - 1:
-			logger.Warn("service metrics groups reached 100% capacity")
+	if entry == nil && mb.entries >= mb.maxSize && mb.other != nil {
+		entry = mb.other
+		// axiomhq/hyerloglog uses metrohash but here we are using
+		// xxhash. Metrohash has better performance but since we are
+		// already calculating xxhash we can use it directly.
+		mb.otherCardinalityEstimator.InsertHash(hash)
+	}
+	if entry != nil {
+		entry.serviceMetrics = serviceMetrics{
+			transactionDuration: entry.transactionDuration + metrics.transactionDuration,
+			transactionCount:    entry.transactionCount + metrics.transactionCount,
+			failureCount:        entry.failureCount + metrics.failureCount,
+			successCount:        entry.successCount + metrics.successCount,
 		}
+		return
+	}
+	if mb.entries >= mb.maxSize {
+		logger.Warnf(`
```

Comment:
Hmm, it will hit this line only once per aggregation interval. The logic will create an overflow bucket per evaluation interval and the logline will only be executed when the bucket is created i.e. the overflow occurs. Once the overflow bucket is created, the entry will be aggregated in [above lines](https://github.com/elastic/apm-server/blob/8f2bc2da8cd062b057ba55e7d5e050dd3807ee7e/x-pack/apm-server/aggregation/servicemetrics/aggregator.go#L288-L305). Thus the code has an implicit rate limited logger were rate = the aggregation interval.